src/pytest/test_external.py

```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2018-2020, Michael P. Howard
+# Copyright (c) 2021-2024, Auburn University
+# Part of azplugins, released under the BSD 3-Clause License.
+
+import hoomd
+from hoomd.azplugins import external
+import pytest
+import numpy
+
+
+class CustomVariant(hoomd.variant.Variant):
+    def __init__(self, z):
+        hoomd.variant.Variant.__init__(self)
+        self.z = z
+
+    def __call__(self, timestep):
+        if timestep <= 1:
+            z = self.z
+        else:
+            z = self.z - 1
+        return z
+
+    def _min(self):
+        return self.z - 1
+
+    def _max(self):
+        return float(self.z)
+
+
+@pytest.fixture
+def custom_variant_fixture():
+    return CustomVariant(z=5.0)
+
+
+@pytest.fixture
+def integrator():
+    ig = hoomd.md.Integrator(dt=0.0)
+    nve = hoomd.md.methods.ConstantVolume(hoomd.filter.All())
+    ig.methods = [nve]
+    return ig
+
+
+@pytest.fixture
+def snap():
+    snap_ = hoomd.Snapshot()
+    if snap_.communicator.rank == 0:
+        snap_.configuration.box = [20, 20, 20, 0, 0, 0]
+        snap_.particles.N = 4
+        snap_.particles.types = ["A", "B"]
+        snap_.particles.position[:] = [
+            [0, 0, 4.6],
+            [0, 0, -5.4],
+            [0, 5.6, 0],
+            [6.6, 0, 0],
+        ]
+        snap_.particles.velocity[:] = [[0, 0, 0]] * 4
+        snap_.particles.typeid[:] = [0, 1, 0, 0]
+    return snap_
+
+
+@pytest.mark.parametrize("cls", [external.SphericalHarmonicBarrier], ids=["Spherical"])
+def test_spherical_harmonic_barrier(
+    simulation_factory, cls, integrator, snap, custom_variant_fixture
+):
+    sim = simulation_factory(snap)
+    sim.operations.integrator = integrator
+
+    barrier = cls(interface=custom_variant_fixture)
+    kA = 50.0
+    dB = 2.0
+    kB = kA * dB**2
+
+    barrier.params["A"] = dict(k=kA, offset=0.1)
+    barrier.params["B"] = dict(k=kB, offset=-0.1)
+
+    sim.operations.add(barrier)
+    #  at first step the interface will not move
+    sim.run(1)
+
+    # Test forces on particles
+    forces = barrier.forces
+    energies = barrier.energies
+    # particle 0 is outside interaction range
+    assert numpy.isclose(energies[0], 0.0)
+    assert numpy.isclose(forces[0][0], 0.0, atol=1e-4)
+    assert numpy.isclose(forces[0][1], 0.0, atol=1e-4)
+    assert numpy.isclose(forces[0][2], 0.0, atol=1e-4)
```

Comment:
You can use `assert_allclose` to check multiple values at once, like:
```suggestion
    assert numpy.testing.assert_allclose(forces[0], [0, 0, 0], atol=1e-4)
```